WcaOnRails/app/webpacker/components/Persons/Badges.jsx

```diff
@@ -0,0 +1,81 @@
+import React from 'react';
+import { Popup } from 'semantic-ui-react';
+import useLoadedData from '../../lib/hooks/useLoadedData';
+import I18n from '../../lib/i18n';
+import { apiV0Urls, delegatesPageUrl, teamsCommitteesPageUrl } from '../../lib/requests/routes.js.erb';
+import { groupTypes, userRolesSortParams as sortParams } from '../../lib/wca-data.js.erb';
+
+// let i18n-tasks know the key is used
+// i18n-tasks-use t('user_groups.group_types.board')
+// i18n-tasks-use t('user_groups.group_types.officers')
+
+function badgeParams(role) {
+  if (role.group.group_type === groupTypes.delegate_regions) {
+    return {
+      roleTitle: I18n.t(`enums.user.role_status.${role.group.group_type}.${role.metadata.status}`),
+      groupTitle: I18n.t(`enums.user.role_status.${role.group.group_type}.${role.metadata.status}`),
+      badgeClass: 'delegate-badge',
+      url: delegatesPageUrl,
+    };
+  }
+  if ([groupTypes.teams_committees, groupTypes.councils].includes(role.group.group_type)) {
+    return {
+      roleTitle: `${role.group.metadata.friendly_id.toUpperCase()} ${I18n.t(`enums.user.role_status.${role.group.group_type}.${role.metadata.status}`)}`,
+      groupTitle: I18n.t(`user_groups.group_types.${role.group.group_type}`),
+      badgeClass: `team-${role.metadata.status.replace('_', '-')}-badge`,
+      url: teamsCommitteesPageUrl,
+    };
+  }
+  if (role.group.group_type === groupTypes.board) {
+    return {
+      roleTitle: role.group.metadata.friendly_id.toUpperCase(),
+      groupTitle: I18n.t(`user_groups.group_types.${role.group.group_type}`),
+      badgeClass: 'team-member-badge',
+      url: teamsCommitteesPageUrl,
+    };
+  }
+  if (role.group.group_type === groupTypes.officers) {
+    return {
+      roleTitle: `${I18n.t(`about.structure.${role.metadata.status}.name`)}`,
+      groupTitle: I18n.t(`user_groups.group_types.${role.group.group_type}`),
+      badgeClass: 'officer-badge',
+      url: teamsCommitteesPageUrl,
+    };
+  }
+  return {};
+}
+
+export default function Badges({ userId }) {
+  const { data } = useLoadedData(apiV0Urls.userRoles.listOfUser(
+    userId,
+    [sortParams.lead, sortParams.eligibleVoter, sortParams.groupTypeRank, sortParams.status, sortParams.groupName].join(','), // Sort params
```

Comment:
Re the other discussion: If you access the keys here, you already know that the result of `sortParams.lead` will be `lead`.

So effectively, it is already hard-coded now because you "magically" know that the sort param `eligibleVoter` exists.
The only good reason to export this constant from Rails to JS is to let JS choose which sort params it should pick **without knowing what sort params exist**. But in the way you're currently using the code, all that's happening is an over-glorified version of "return a string that is called exactly like the variable name which I chose"

Reply:
Okay got it, makes sense. What I was thinking is, if I do something like `sortParams.nonExistantParam`, it will show some error, but we don't get error in frontend via lint or anything, so it doesn't makes sense to have it. I've now removed it.

Reply:
That is a very good thought and I appreciate the idea. However our infrastructure doesn't support it very well right now and because you tried so hard to squeeze it into a non-typed JS stack you ended up with a super-complicated implementation that took me 6 or so GitHub comments to figure out what's even going on.

Reply:
Super-nice idea in theory, but please don't break a leg with implementing it, especially in a legacy tech stack that has lots of technical debt.

If your implementation contains `each_with_hash`, it's already a very high chance that something is not supported in the way you want. If you even have to pass lambda params `|(key, value), new_hash|` into the `each_with_hash`, then you know there's **definitely** an un-supported feature that you're trying too hard to figure out ;)

Reply:
Got it, thanks a lot for explaining. 🙂